### ISSUE TITLE
Output is different after first run and after "resume" run in particular cases

### ISSUE THREAD

**Issue by fredericlemoine** [nextflow-io/nextflow#135]
The output of the following pipeline is different after the first run, and after a "resume" run.

```groovy
inchannel = Channel.from(1,1,1,1,1,2,2,2,2,2)
process random{
        input:
        val(i) from inchannel

        output:
        file("sample_${i}.txt") into randout

        shell:
        """
        echo !{i} \${RANDOM} > sample_!{i}.txt
        """
}
randfile = randout.collectFile(name: 'outfile.txt')
randfile.subscribe{
        println it
}
```
In the original run, the final output file contains:
```
2 4828
1 17606
1 30810
2 14822
2 284
2 6645
1 24875
1 768
2 26209
1 15789
```

In the "resume" run the final output file contains:
```
1 17606
1 17606
1 17606
1 17606
1 17606
2 284
2 284
2 284
2 284
2 284
```

What I understand is that since several process instances have the same output filename, it does not pose any problems in the first run. 
However, in "resume" mode, the "collectfile" takes 5 times the same output file from the "random" process for value "1" and 5 times the same output file from the "random" process for the value "2".

I hope my description helps,

Best

**Comment by pditommaso** [comment]
Thanks a lot for the detailed report. The Nextflow caching mechanism assumes that processes have deterministic-functional behaviour i.e. a process output the same values when the same inputs are provided. In your example the shell `$RANDOM` variable breaks this contract producing this weird result. 

More in detail what happens is that when a process is executed nextflow creates an hash key indexing the process script and the provided input values. When the pipeline execution is *resumed*, it picks up the an existing output for which exists a matching cache key. 

Since in your script the `$RANDOM` variable is resolved at shell level what happens is that only two valid cache keys are produced because the script is the same the cache key are created. 

In conclusion if process need to use a random variable you can either: 
- disable the caching mechanism for that process adding [cache false](http://www.nextflow.io/docs/latest/process.html#cache) directive  
- guarantee different input values for each run of that process. 

Does this solve your issue?

**Comment by pditommaso** [comment]
One more thing: you can avoid to escape the `$` character using a `'''` multiline string in place of a `"""`. For example: 

        shell:
        '''
        echo !{i} ${RANDOM} > sample_!{i}.txt
        '''

**Comment by fredericlemoine** [comment]
Thank you for your clear answer.
I think I will have to give different input values in my pipeline then.
It solves the issue.

btw, for escaping the `$`, thanks for the tip.
